### ISSUE TITLE
Certificates with wildcards are not handled correctly

### ISSUE THREAD

**Issue by jbevemyr** [ninenines/gun#270]
The customize_hostname_check needs to be set to a fun that performs
https style hostname checks. Different protocols using tls wants
different handling of hostname matching in certificates, ie https and
ldap differ in how wildcards should be handled.

By default wildcards are not handled correctly for https, ie
connecting to https://aus.auth0.com does not work when

tls_opts => [{verify,verify_peer},
             {cacertfile, "/etc/ssl/certs/ca-certificates.crt"}]

it also needs
{customize_hostname_check, [{match_fun, CustomFun}]}  where
CustomFun = public_key:pkix_verify_hostname_match_fun(https)

**Comment by jbevemyr** [comment]
Yes, the fix has not been incorporated but you can use my fork if you like or pester the maintainer to include the fix :-). The diff is very small.  https://github.com/jbevemyr/gun
There are two commits in my fork that solves the problem.

https://github.com/jbevemyr/gun/commit/fdd196c6ed0a1c5d5cd80d7ee2495c23a2114b55
https://github.com/jbevemyr/gun/commit/6d0ee4feca139c1b8d30f4873f7d9b5606f2f58f

**Comment by mfeckie** [comment]
@essen That sounds great, can you give me a hint on how to do that?

**Comment by jbevemyr** [comment]
Something like this

CustomFun = public_key:pkix_verify_hostname_match_fun(https),
TransOpts = [{customize_hostname_check, [{match_fun, CustomFun}]}],
ConnectionOpts = #{transport => tls, tls_opts => TransOpts},
gun:open(Host, Port, ConnectionOpts)

If you are using Erlang.

**Comment by essen** [comment]
Does anything need to be done in Gun with regard to this? Can we close this?

**Comment by jbevemyr** [comment]
I think the gun should be changed to use the public_key:pkix_verify_hostname_match_fun(https) fun to check hostnames since that is the expected behaviour when using https.

**Comment by essen** [comment]
Can that be set without setting any other TLS option? Considering Gun currently does not verify certificates and I don't think this will change in 2.0.

**Comment by jbevemyr** [comment]
Right, that is true. Don't know if it can be set separately.

**Comment by essen** [comment]
Starting with Gun 2.2 the `customize_hostname_check` function should be set by default. That plus the previously added `cacerts` configuration by default (using `public_key:cacerts_get/0`) means that as far as certificates are concerned, the default should now be fully compliant. Users may want to keep configuring some options such as `depth` though.

Closing, thanks!!
